odl/discr/lp_discr.py

```diff
@@ -221,18 +229,88 @@ def element(self, inp=None):
                         ''.format(arr.shape, self.shape))
             arr = arr.ravel(order=self.order)
             return self.element_type(self, self.dspace.element(arr))
-        except TypeError as err:
+        except (TypeError, IndexError) as err:
             if str(err.args[0]).startswith('output contains points outside'):
                 raise err
             else:
                 raise_from(TypeError('unable to create an element of {} from '
-                                     '{!r}.'.format(self, inp)), err)
+                                     '{!r}: {}'.format(self, inp, err)), err)
 
     @property
     def interp(self):
         """Interpolation type of this discretization."""
         return self._interp
 
+    def as_complex_space(self, dtype=None):
```

Comment:
But then, the `astype` is trivial, at least for `FnBase` spaces:
```python
def astype(self, dtype):
    return type(self)(self.size, dtype=dtype, weight=self.weighting)
```
It's more complicated for, e.g., `DiscreteLp` so it may well be worth a method. On the other hand, the nice thing with the real <-> complex translation is that you don't have to manually type all those if's and else's for the 4 possible cases (r2r, r2c, c2r, c2c) in all situations it pops up. I'd argue that this is worth considering as an addition (when you deal with Fourier transforms, these conversions are *everywhere*).